crates/bevy_ecs/src/observer/emit_trigger.rs

```diff
@@ -0,0 +1,160 @@
+use crate::{
+    component::ComponentId,
+    entity::Entity,
+    event::Event,
+    world::{Command, DeferredWorld, World},
+};
+
+/// A [`Command`] that emits a given trigger for a given set of targets.
+pub struct TriggerEvent<E, Targets: TriggerTargets = ()> {
+    /// The event to trigger.
+    pub event: E,
+
+    /// The targets to trigger the event for.
+    pub targets: Targets,
+}
+
+impl<E: Event, Targets: TriggerTargets> Command for TriggerEvent<E, Targets> {
+    fn apply(mut self, world: &mut World) {
+        let event_type = world.init_component::<E>();
+        trigger_event(world, event_type, &mut self.event, self.targets);
+    }
+}
+
+/// Emit a trigger for a dynamic component id. This is unsafe and must be verified manually.
+pub struct EmitDynamicTrigger<T, Targets: TriggerTargets = ()> {
+    event_type: ComponentId,
+    event_data: T,
+    targets: Targets,
+}
+
+impl<E, Targets: TriggerTargets> EmitDynamicTrigger<E, Targets> {
+    /// Sets the event type of the resulting trigger, used for dynamic triggers
+    /// # Safety
+    /// Caller must ensure that the component associated with `event_type` is accessible as E
+    pub unsafe fn new_with_id(event_type: ComponentId, event_data: E, targets: Targets) -> Self {
+        Self {
+            event_type,
+            event_data,
+            targets,
+        }
+    }
+}
+
+impl<E: Event, Targets: TriggerTargets> Command for EmitDynamicTrigger<E, Targets> {
+    fn apply(mut self, world: &mut World) {
+        trigger_event(world, self.event_type, &mut self.event_data, self.targets);
+    }
+}
+
+#[inline]
+fn trigger_event<E, Targets: TriggerTargets>(
+    world: &mut World,
+    event_type: ComponentId,
+    event_data: &mut E,
+    targets: Targets,
+) {
+    let mut world = DeferredWorld::from(world);
+    if targets.entities().len() == 0 {
+        // SAFETY: T is accessible as the type represented by self.trigger, ensured in `Self::new`
+        unsafe {
+            world.trigger_observers_with_data(
+                event_type,
+                Entity::PLACEHOLDER,
+                targets.components(),
+                event_data,
+            );
+        };
+    } else {
+        for target in targets.entities() {
+            // SAFETY: T is accessible as the type represented by self.trigger, ensured in `Self::new`
+            unsafe {
+                world.trigger_observers_with_data(
+                    event_type,
+                    target,
+                    targets.components(),
+                    event_data,
+                );
+            };
+        }
+    }
+}
+
+/// Represents a collection of targets, which can be of type [`Entity`] or [`ComponentId`].
```

Comment:
This needs a link back to observers in some way.

It should also have a bit of an explanation about "what does it mean for something to be a target".